### ISSUE TITLE
tf.keras NaN loss when using multiple GPUs

### ISSUE THREAD

**Issue by rodyt** [tensorflow/tensorflow#36224]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Pip
- TensorFlow version (use command below): 2.0
- Python version: 3.7.3
- Bazel version (if compiling from source): no
- GCC/Compiler version (if compiling from source): no
- CUDA/cuDNN version: 10
- GPU model and memory: 2x NVIDIA GTX 1080 Ti 11GB
- Driver Version: 440.33.01

I am currently using Tensorflow 2.0 (Python) and the` tf.keras` library to train a CNN. However, I am encountering an issue when I try to train my model by calling `model.fit()`. 

After I begin training, the loss is normal for 1 ~ 2 steps for the first epoch. But after that, it suddenly becomes **NaN** loss.

This issue only happens when using multiple GPUs. The code I'm using works perfectly fine on a single GPU. I have wrapped all of my code inside the scope of a `tf.distribute.MirroredStrategy ` using `with strategy.scope():`. I am feeding my network with data from a `tf.data.Dataset` (though this error occurs regardless of the data format).

I then ran some tests:

1) I tried to replace the data in my dataset with random number, but the loss still went to **NaN**.

2) I also tried feeding the numpy arrays directly to `.fit()`, but that didn't solve the issue.

3) I tried using different optimizers (Adam, RMSprop, SGD), batch sizes (4, 8, 16, 32), and learning rates, none of which helped to solve this problem.

4) I swapped out my network for a simple Multi-layer Perceptron, but the error persisted.

This doesn't appear to be an OOM issue, since the data is relatively small and running `watch -n0.1 nvidia-smi` reveals that memory usage never exceeds 30% on either of my GPUs. There doesn't seem to be any warning or error in the console output that might hint at the issue either.

Any help is appreciated

**Comment by Saduf2019** [comment]
Could you please provide us with simple standalone code to reproduce the issue in our environment, Thanks.

**Comment by rodyt** [comment]
Hi @Saduf2019 ,

Thanks for responding. I am able to reproduce the error with the sample MNIST code provided on the Tensorflow website:
https://www.tensorflow.org/tutorials/distribute/keras

This is the sample code from the tutorial that I tried to run:
```

import tensorflow_datasets as tfds
import tensorflow as tf
tfds.disable_progress_bar()
import os

datasets, info = tfds.load(name='mnist', with_info=True, as_supervised=True)
mnist_train, mnist_test = datasets['train'], datasets['test']

num_train_examples = info.splits['train'].num_examples
num_test_examples = info.splits['test'].num_examples

BUFFER_SIZE = 10000

BATCH_SIZE_PER_REPLICA = 64
BATCH_SIZE = BATCH_SIZE_PER_REPLICA * 2

def scale(image, label):
    image = tf.cast(image, tf.float32)
    image /= 255

    return image, label

train_dataset = mnist_train.map(scale).cache().shuffle(BUFFER_SIZE).batch(BATCH_SIZE)
eval_dataset = mnist_test.map(scale).batch(BATCH_SIZE)

strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, 3, activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D(),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(64, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
    ])
    
    model.compile(loss='sparse_categorical_crossentropy',
                optimizer=tf.keras.optimizers.Adam(),
                metrics=['accuracy'])
    model.fit(
        train_dataset, 
        epochs=12
    )

```
This is the output when I run `.fit` using `MirroredStrategy`:

> Epoch 1/12
>
> INFO:tensorflow:batch_all_reduce: 6 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
> INFO:tensorflow:batch_all_reduce: 6 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
> INFO:tensorflow:batch_all_reduce: 6 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
> INFO:tensorflow:batch_all_reduce: 6 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
>      16/Unknown - 7s 423ms/step - loss: **nan** - accuracy: 0.0932

This is the output of the model when I run _without_ `MirroredStrategy`:

> 
> Epoch 1/12
> 469/469 [==============================] - 2s 4ms/step - loss: 0.0480 - accuracy: 0.9853
> Epoch 2/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0361 - accuracy: 0.9892
> Epoch 3/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0302 - accuracy: 0.9908
> Epoch 4/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0249 - accuracy: 0.9922
> Epoch 5/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0197 - accuracy: 0.9944
> Epoch 6/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0163 - accuracy: 0.9950
> Epoch 7/12
> 469/469 [==============================] - 1s 3ms/step - loss: 0.0122 - accuracy: 0.9966
> Epoch 8/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0102 - accuracy: 0.9970
> Epoch 9/12
> 469/469 [==============================] - 1s 3ms/step - loss: 0.0082 - accuracy: 0.9977
> Epoch 10/12
> 469/469 [==============================] - 1s 3ms/step - loss: 0.0087 - accuracy: 0.9975
> Epoch 11/12
> 469/469 [==============================] - 2s 3ms/step - loss: 0.0060 - accuracy: 0.9983
> Epoch 12/12
> 469/469 [==============================] - 1s 3ms/step - loss: 0.0051 - accuracy: 0.9984

**Comment by anj-s** [comment]
I am unable to repro the issue in a colab using the standalone snippet you provided.  Can you try using [tf.debugging.enabled_check_numerics](https://www.tensorflow.org/api_docs/python/tf/debugging/enable_check_numerics) to help pinpoint the op that is causing this?

**Comment by rodyt** [comment]
I'm sorry, I don't have TF 2.1. What is the alternative to this?

**Comment by bardenthenry** [comment]
> I'm sorry, I don't have TF 2.1. What is the alternative to this?

Well, I can tell you that this problem have not been sloved at tensorflow 2.1....

**Comment by chrisffm** [comment]
Same here with TF 2.1, have been using the same example and also get the NaN-Values while the training on 2 gpu's is much slower than on one

**Comment by rodyt** [comment]
Is there any other way I can provide debug information on TF 2.0? It's incredibly annoying to not be able to use all my GPUs. @anj-s  @jvishnuvardhan

**Comment by VyBui** [comment]
@rodyt I have got them same issue while using Keras in multiple GPUs. Have you solved your problem? 
![Screenshot from 2020-02-06 16-35-29](https://user-images.githubusercontent.com/15938335/73924270-ca3b5300-48fe-11ea-8cdf-0f55fdc91137.png)

**Comment by guptapriya** [comment]
@rodyt we are unable to reproduce this problem with the code snippet you provided. Is it possible that there is a driver/hardware issue with one of the GPUs? 

@VyBui please file a separate ticket for your issues - there could be many different reasons for getting a NaN. If you file a separate ticket with your code to repro, setup etc, that can help us look into that issue.

**Comment by saikumarchalla** [comment]
@rodyt  Closing this issue as we are unable to reproduce this problem with the code snippet you provided.Thanks!

**Comment by mepster** [comment]
This issue still exists in tf / keras 2.8.0, so it shouldn't really have been closed.

There is a workaround described here:
https://stackoverflow.com/questions/54283937/training-on-multiple-gpus-causes-nan-validation-errors-in-keras

Quoting from that link: "As far as I can (heuristically) tell, when doing distributed training/evaluation, the number of elements in the dataset must be evenly divisible by the batch size and number of GPUs. That is, nelements / ngpus / batch_size == 0. If that is not the case, then empty batches will be passed to the loss function, which, depending on the loss function, may inject NaN losses into the aggregator."

This explanation matches what I see and his workaround works for me. But it should be fixed in tf: empty batches should not be created.
